lib/engine/game/g_18_fl/step/token.rb

```diff
@@ -42,16 +48,36 @@ def process_hex_token(action)
             token = action.token
 
             raise GameError, "#{hex.id} is not a town" if hex.tile.towns.empty?
-            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex,
-                                                                                                          entity)
+            raise GameError, "#{entity.name} already has a hotel in #{hex.tile.location_name}" if tokened(hex, entity)
+
+            cost = action.cost # We are using token_cost_override
+            raise GameError, "#{entity.name} cannot afford "\
+                  "#{@game.format_currency(cost)} cost to lay hotel" if cost > entity.cash
 
-            @game.log << "#{entity.name} places a hotel on #{hex.name}"
+            @game.log << "#{entity.name} places a hotel on #{hex.name} for #{@game.format_currency(cost)}"
+            entity.spend(cost, @game.bank)
 
             entity.tokens.delete(token)
             hex.tile.icons << Part::Icon.new("../logos/18_fl/#{entity.id}")
             pass!
           end
 
+          def token_cost_override(entity, city_hex, _slot, token, _tokener)
+            node = city_hex.respond_to?(:city?) ? city_hex : city_hex.tile.towns.first
+            min_distance = INFINITE_DISTANCE
+            goal_cities = entity.tokens.select(&:city).map(&:city)
+
+            node.walk(corporation: entity) do |path, visited_paths|
+              if goal_cities.include?(path.city)
+                # minus one to account for the tokened hex getting included in the count
+                distance = visited_paths.select { |_k, v| v }.keys.map(&:hex).uniq.size - 1
```

Comment:
visited_paths.uniq { |k, _| k.hex }.size - 1

v is always true

Reply:
Thanks, I 've updated this